Adds protocol compliance tests to the Http4s module

yisraelU opened the pull request:
This replaces the pizza spec with the protocol compliance tests using alloy-protocol-tests
In addition, the tests for ComplianceTests module were removed as they were not testing the module , merely using it
Will remove additional components once I confirm what is no longer necessary

yisraelU committed:
replaces pizza spec with the protocol compliance tests using alloy-protocol-tests
Will remove additional componenets once I confirm what is no longer necessary

Baccata commented on modules/http4s/test/src/smithy4s/http4s/Http4sPizzaSpec.scala:
```diff
@@ -1,43 +0,0 @@
-import org.http4s.implicits._
-import smithy4s.example.PizzaAdminService
-
-object Http4sPizzaSpec extends smithy4s.tests.PizzaSpec {
```
Unless I'm mistaken, I believe there might be tests in PizzaSpec that cannot be checked using compliance tests, that have to do with the `mapError` combinator. So I'd expect this suite to exist, but I'd expect a lot of tests to be eliminated from the `PizzaSpec` trait.

yisraelU replied:
Ok Then I will go through one by one and prune

yisraelU replied:
@Baccata I think most of the pizza spec will still be required , not just error transformation as the tests are testing specifically the http4s implementation not the protocol, i.e. a not found path returns a 404 or certain requests produce certain responses. Perhaps we should just leave it as is rather then micro pruning it

kubukoz replied:
> testing specifically the http4s implementation not the protocol

isn't that what all the protocol tests do? Test the interpreter for conforming to the protocol.

yisraelU replied:
no, specifically aws protocol tests , test Smithy-Model -> HTTP and vise versa , not HTTP -> HTTP

Baccata replied:
Here are the tests I think can be removed in favour of smithy-defined tests : 

* https://github.com/disneystreaming/smithy4s/blob/series/0.17/modules/tests/src/smithy4s/tests/PizzaSpec.scala#L405-L433
* https://github.com/disneystreaming/smithy4s/blob/series/0.17/modules/tests/src/smithy4s/tests/PizzaSpec.scala#L70-L83
* https://github.com/disneystreaming/smithy4s/blob/series/0.17/modules/tests/src/smithy4s/tests/PizzaSpec.scala#L85-L117
* https://github.com/disneystreaming/smithy4s/blob/series/0.17/modules/tests/src/smithy4s/tests/PizzaSpec.scala#L141-L161
* https://github.com/disneystreaming/smithy4s/blob/series/0.17/modules/tests/src/smithy4s/tests/PizzaSpec.scala#L335-L357
* https://github.com/disneystreaming/smithy4s/blob/series/0.17/modules/tests/src/smithy4s/tests/PizzaSpec.scala#L384-L403
* https://github.com/disneystreaming/smithy4s/blob/series/0.17/modules/tests/src/smithy4s/tests/PizzaSpec.scala#L510-L543

As they test the implementation of the protocol. The fact that they might be testing it differently to what we'd be doing in smithy-written tests obviously doesn't matter that much.

kubukoz commented on build.sbt:
```diff
@@ -645,11 +646,13 @@ lazy val http4s = projectMatrix
     Test / smithySpecs := Seq(
       (ThisBuild / baseDirectory).value / "sampleSpecs" / "hello.smithy"
     ),
+     Test / smithy4sSkip := Seq("openapi"),
```
**issue:** this file needs formatting

kubukoz commented on project/Dependencies.scala:
```diff
@@ -26,9 +26,10 @@ object Dependencies {
+    val version = "0.1.8"
     val core = org % "alloy-core" % version
     val openapi = org %% "alloy-openapi" % version
+    val `protocol-tests` = org % "alloy-protocol-tests" % version
```
**suggestion:** I'd just call it `protocolTests` to avoid the backticks.

yisraelU committed:
reverted deleted http4s specific pizza spec
removed some tests from PizzaSpec as they are no longer necessary due to coverage by compliance tests

yisraelU commented:
Current failure is due to scala 3 native failing when running the Protocol Compliance Tests , which is as of yet an unsolved problem see #680

Baccata committed:
Formatting sbt build

Baccata committed:
showcasing the issue

Baccata commented:
@yisraelU the problem is that the regex you used [there](https://github.com/disneystreaming/smithy4s/blob/series/0.17/modules/compliance-tests/src/smithy4s/compliancetests/internals/package.scala#L85-L86) is probably not working in native c/c++.

I suck at regexes, but I suggest you either go to the scala discord, channel scala-native, and ask for help to port the regex in a way that will work on JVM and Native alike, or use an online tool like https://regex101.com/ to experiment in both java and go to find a regex that will work in both. If it works in go chances are it'll work in scala-native too.

Baccata commented:
You can also experiment with the following scala-cli script to test regexes on native : 

```scala
//> using platform "native"

@main
def main() = {
  val commaNotBetweenQuotes = ",(?=([^\"]*\"[^\"]*\")*[^\"]*$)"
  """foo""".split(commaNotBetweenQuotes)
}
```

you'll see that in the current state it throws

yisraelU commented:
@Baccata great , will do

kubukoz commented on modules/http4s/test/src/compliancetests/ProtocolComplianceTest.scala:
```diff
@@ -22,11 +22,11 @@ import org.http4s._
 import org.http4s.client.Client
 import smithy4s.kinds.FunctorAlgebra
 import smithy4s.Service
-import smithy4s.example.test._
```
idk about everyone else, but my github gave up trying to syntax highlight this file.

kubukoz commented on modules/tests/src/smithy4s/tests/PizzaSpec.scala:
```diff
@@ -67,55 +61,6 @@ abstract class PizzaSpec
     "price" -> Json.fromFloatOrNull(9.0f)
   )
 
-  val positiveEmptyLabel = """|Happy path:
```
Are there any tests in PizzaClientSpec that we could similarly remove?

yisraelU committed:
replace regex with stateful foldLeft for header parsing to provide support for scala native

yisraelU committed:
Merge branch 'series/0.17' into compliance-tests-http4s

yisraelU committed:
alloyVersion rename

yisraelU committed:
alloyVersion rename

yisraelU commented:
replaced the regex with a foldLeft to parse multi headers

lewisjkl commented on modules/compliance-tests/src/smithy4s/compliancetests/internals/package.scala:
```diff
@@ -82,15 +82,33 @@ package object internals {
+  }
 
+  // splits string into List utilizing a comma that is not between quotes as quotes can be used to escape commas in headers
+  private def parseList(s: String): List[String] = {
+    s.foldLeft((Chain.empty[String], 0, 0, false)) {
+      case ((acc, begin, end, quote), elem) =>
+        elem match {
+          case '"' => (acc, begin, end + 1, !quote)
+          case ',' if !quote && begin < end =>
+            (acc :+ s.substring(begin, end), end + 1, end + 1, quote)
+          case c if c.isWhitespace && begin == end =>
+            (acc, begin + 1, end + 1, quote)
+          case _ =>
+            if (s.length == end + 1)
+              (acc :+ s.substring(begin, end + 1), end + 1, end + 1, quote)
+            else {
+              (acc, begin, end + 1, quote)
+            }
+        }
+    }._1
+      .toList
```
It's a bit unclear to me what some of these cases are doing.. I might just be dense, but I think some comment  explaining more what edge cases you are handling here would be helpful

yisraelU replied:
you are definitely not dense.
Ill add clear (hopefully) explanations

lewisjkl replied:
Looks good, thanks for adding 👍

yisraelU committed:
more explicit documentation, plus handle a string that starts with comma

yisraelU committed:
formatting

yisraelU commented:
Will add tests for parseList function tomorrow

yisraelU committed:
make parseList package private instead of private

yisraelU committed:
test for parseList function added to split headers properly

Baccata commented on modules/compliance-tests/test/src/ComplianceUtilsSpec.scala:
```diff
@@ -0,0 +1,53 @@
+
+import smithy4s.compliancetests.internals.parseList
+import weaver._
+object ComplianceUtilsSpec extends FunSuite {
```
this is great, thank you :)
